tsl/src/remote/prepared_statement_fetcher.c

```diff
@@ -0,0 +1,402 @@
+		tuplefactory_reset_mctx(fetcher->state.tf);
+
+		fetcher->state.num_tuples = i;
+		fetcher->state.next_tuple_idx = 0;
+		fetcher->state.batch_count++;
+
+		if (fetcher->state.eof)
+		{
+			fetcher->state.data_req = NULL;
+		}
+	}
+	PG_CATCH();
+	{
+		if (NULL != fetcher->state.data_req)
+		{
+			fetcher->state.data_req = NULL;
+		}
+
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+
+	MemoryContextSwitchTo(oldcontext);
+
+	return fetcher->state.num_tuples;
+}
+
+static int
+prepared_statement_fetcher_fetch_data(DataFetcher *df)
+{
+	PreparedStatementFetcher *fetcher = cast_fetcher(PreparedStatementFetcher, df);
+
+	if (fetcher->state.eof)
+		return 0;
+
+	if (!fetcher->state.open)
+		prepared_statement_fetcher_send_fetch_request(df);
+
+	return prepared_statement_fetcher_complete(fetcher);
+}
+
+static void
+prepared_statement_fetcher_store_tuple(DataFetcher *df, int row, TupleTableSlot *slot)
+{
+	PreparedStatementFetcher *fetcher = cast_fetcher(PreparedStatementFetcher, df);
+
+	ExecClearTuple(slot);
+
+	if (row >= df->num_tuples)
+	{
+		if (df->eof || df->funcs->fetch_data(df) == 0)
+		{
+			return;
+		}
+
+		row = 0;
+		Assert(row == df->next_tuple_idx);
+	}
+
+	Assert(fetcher->batch_values != NULL);
+	Assert(fetcher->batch_nulls != NULL);
+	Assert(row >= 0 && row < df->num_tuples);
+
+	const int nattrs = tuplefactory_get_nattrs(fetcher->state.tf);
+	slot->tts_values = &fetcher->batch_values[nattrs * row];
+	slot->tts_isnull = &fetcher->batch_nulls[nattrs * row];
+	ExecStoreVirtualTuple(slot);
+}
+
+static void
+prepared_statement_fetcher_store_next_tuple(DataFetcher *df, TupleTableSlot *slot)
+{
+	prepared_statement_fetcher_store_tuple(df, df->next_tuple_idx, slot);
+
+	if (!TupIsNull(slot))
+		df->next_tuple_idx++;
+
+	Assert(df->next_tuple_idx <= df->num_tuples);
+}
+
+DataFetcher *
+prepared_statement_fetcher_create_for_scan(TSConnection *conn, const char *stmt, StmtParams *params,
+										   TupleFactory *tf)
+{
+	PreparedStatementFetcher *fetcher = palloc0(sizeof(PreparedStatementFetcher));
+
+	data_fetcher_init(&fetcher->state, conn, stmt, params, tf);
+	fetcher->state.type = PreparedStatementFetcherType;
+	fetcher->state.funcs = &funcs;
+
+	PGconn *pg_conn = remote_connection_get_pg_conn(conn);
+	if (remote_connection_get_status(conn) != CONN_IDLE)
+	{
+		elog(ERROR,
+			 "unexpected activity on data node connection when creating the row-by-row fetcher");
+	}
+
+	/*
+	 * Force using the generic plan for each execution of the data node query,
+	 * because it would be very expensive and pointless to replan it for each
+	 * subsequent parameter value. The cleanup will be done by the remote
+	 * connection layer when it initializes a remote connection. This way we
+	 * avoid an extra network request on an erroneous exit or
+	 * cancellation, which would be both prone to cascading errors and
+	 * time-consuming.
+	 */
+	PGresult *res = remote_connection_exec(conn, "SET plan_cache_mode = 'force_generic_plan'");
```

Comment:
When we execute it, is it on first connection to the DN? seems like expensive to do

Reply:
This code runs only when the fetcher is created, basically once per top-level query that the user runs on the AN.

Reply:
So we send an additional command for each query each time? Does it makes sense to keep a state per connection and send it only if necessary to reset. We are doing something similar already for connection options